pkg/utils/report/results.go

```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
-						"version":  pss.Version,
-						"controls": strings.Join(controls, ","),
-					}
-				}
-			}
-			if result.Result == "fail" && !result.Scored {
-				result.Result = "warn"
-			}
-			results = append(results, result)
+func addProperty(k, v string, result *policyreportv1alpha2.PolicyReportResult) {
+	if result.Properties == nil {
+		result.Properties = map[string]string{}
+	}
+
+	result.Properties[k] = v
+}
+
+func ToPolicyReportResult(policyType engineapi.PolicyType, policyName string, ruleResult engineapi.RuleResponse, annotations map[string]string, resource *corev1.ObjectReference) policyreportv1alpha2.PolicyReportResult {
+	result := policyreportv1alpha2.PolicyReportResult{
+		Source:  kyverno.ValueKyvernoApp,
+		Policy:  policyName,
+		Rule:    ruleResult.Name(),
+		Message: ruleResult.Message(),
+		Result:  toPolicyResult(ruleResult.Status()),
+		Scored:  annotations[kyverno.AnnotationPolicyScored] != "false",
+		Timestamp: metav1.Timestamp{
+			Seconds: time.Now().Unix(),
+		},
+		Category: annotations[kyverno.AnnotationPolicyCategory],
+		Severity: SeverityFromString(annotations[kyverno.AnnotationPolicySeverity]),
+	}
+	if result.Result == "fail" && !result.Scored {
+		result.Result = "warn"
+	}
+	if resource != nil {
+		result.Resources = []corev1.ObjectReference{
+			*resource,
 		}
-	} else {
-		for _, ruleResult := range response.PolicyResponse.Rules {
-			result := policyreportv1alpha2.PolicyReportResult{
-				Source:  "ValidatingAdmissionPolicy",
-				Policy:  ruleResult.Name(),
-				Message: ruleResult.Message(),
-				Result:  toPolicyResult(ruleResult.Status()),
-				Timestamp: metav1.Timestamp{
-					Seconds: time.Now().Unix(),
-				},
-			}
-			if ruleResult.ValidatingAdmissionPolicyBinding() != nil {
-				result.Properties = map[string]string{
-					"binding": ruleResult.ValidatingAdmissionPolicyBinding().Name,
-				}
+	}
+	if ruleResult.Exception() != nil {
+		addProperty("exception", ruleResult.Exception().Name, &result)
+	}
+	pss := ruleResult.PodSecurityChecks()
+	if pss != nil && len(pss.Checks) > 0 {
+		var controls []string
+		for _, check := range pss.Checks {
+			if !check.CheckResult.Allowed {
+				controls = append(controls, check.ID)
 			}
-			results = append(results, result)
 		}
+		if len(controls) > 0 {
+			sort.Strings(controls)
+			addProperty("standard", string(pss.Level), &result)
+			addProperty("version", pss.Version, &result)
+			addProperty("controls", strings.Join(controls, ","), &result)
+		}
+	}
+	if policyType == engineapi.ValidatingAdmissionPolicyType {
+		result.Source = "ValidatingAdmissionPolicy"
+		if ruleResult.ValidatingAdmissionPolicyBinding() != nil {
+			addProperty("binding", ruleResult.ValidatingAdmissionPolicyBinding().Name, &result)
+		}
+	}
+
+	return result
+}
+
+func EngineResponseToReportResults(response engineapi.EngineResponse) []policyreportv1alpha2.PolicyReportResult {
+	pol := response.Policy()
+	policyName, _ := cache.MetaNamespaceKeyFunc(pol.AsKyvernoPolicy())
```

Comment:
This is the issue. You have to check the policy type and according to it, get either the Kyverno policy name or the VAP name because now, you always get the Kyverno policy name. Moreover, the VAP name can be extracted from the ruleResponse, there is no need to get it from the cache.

Reply:
Thanks @MariamFahmy98! 

@coder12git - can you add me as a collaborator to your repo, so I can push changes?

Reply:
@MariamFahmy98 - would a `pol.GetName()` not work here for both Kyverno policies and ValidatingAdmissionPolicy types?

Reply:
> @coder12git - can you add me as a collaborator to your repo, so I can push changes?

@JimBugwadia i have added you as a collaborator.

Reply:
> @MariamFahmy98 - would a `pol.GetName()` not work here for both Kyverno policies and ValidatingAdmissionPolicy types?

It will work.